clippy_lints/src/methods/mod.rs

```diff
@@ -927,6 +928,33 @@ declare_clippy_lint! {
     "using `.skip(x).next()` on an iterator"
 }
 
+declare_clippy_lint! {
+    /// **What it does:** Checks for use of `obj.method().for_each(closure)` if obj doesn't
+    /// implelement `Iterator` and `method()` returns `Impl Iterator` type.
+    ///
+    /// **Why is this bad?** Excessive use of `for_each` reduces redability, using `for` loop is
+    /// clearer and more concise.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// let v = vec![0, 1, 2];
+    /// v.iter().for_each(|elem| println!("{}", elem));
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let v = vec![0, 1, 2];
+    /// for elem in v.iter() {
+    ///     println!("{}", elem);
+    /// }
+    /// ```
+    pub EXCESSIVE_FOR_EACH,
+    style,
+    "using `.iter().for_each(|x| {..})` when using `for` loop would work instead"
+}
```

Comment:
should this really warn by default?
IMO there is nothing "wrong" with `.iter().for_each()`.
`.iter().for_each()` also enables us to parallelize the code using `par_iter().for_each()` which is not the case for a traditional `for` loop.

Reply:
Indeed, so how about using `restriction` instead?

> .iter().for_each() also enables us to parallelize the code using par_iter().for_each() which is not the case for a traditional for loop.

In that case, this lint isn't  triggered.  `obj.method().for_each(closure)`  triggers this lint iff `obj.method()` is NOT a `Iterator` method and type of `obj` is of `std::collection` or `vec` or `slice`, or `array`.

Reply:
I think restriction might be better fitting :+1: 

I ran clippy with that lint on a couple of crates
https://github.com/rust-lang/rust-clippy/blob/master/clippy_dev/lintcheck_crates.toml
and it did not seem to trigger a single time which was surprising to me :thinking:  (`cargo dev-lintcheck`)